ruby/computer_science/space_complexity.md

```diff
@@ -0,0 +1,141 @@
+
+~~~ruby
+def sum_hash_values(hash)
+  copy_hash = hash.dup
+  sum = 0
+  copy_hash.each do |_, value|
+    sum += value
+  end
+  return sum
+end
+~~~
+
+Here as the hash size increases, the space it uses grows in a linear way.
+
+#### Other complexities
+
+As we've stated, many data structures share O(N) space complexity and therefore you won't write many algorithms with a space complexity that differs
+
+You do find some recursive functions that may have a different space complexity and some sorting algorithms. You normally won't have much reason to consider anything else though.
+
+In the last lesson one of the assignments was a link to the [Big-O cheat sheet](https://www.bigocheatsheet.com/). If you take another look at it now you may have a better appreciation for just how amazing it is as a reference for space and time complexity. If you scroll down to the data structures and then the sorting algorithms section you'll see it gives you the time and space complexities. Notice just how many are O(N), especially for data structures. Many sorting algorithms have just O(1) space complexity, something to keep in mind as you come across different sorting algorithms during your learning.
+
+That's why we won't be diving into examples for other Big O notations with space complexity. We'd have to come up with convuluted examples that didn't represent most code you'll write. If you do come across a good real world example in your own code then do let us know and we may consider adding it here for others to consider.
+
+#### Other considerations
+
+One of the common areas that causes confusion when considering space complexity is what constitutes using space in the context of an algorithm. In an earlier example we wrote methods that duplicated an array and hash argument. We did that to be explicit. But what if we'd written the method as:
+
+~~~ruby
+def sum_arr(arr)
+  sum = 0
+  arr.each do |number|
+    sum += number
+  end
+  return sum
+end
+~~~
+
+When a data structure is passed in as the argument, especially for languages that pass arrays by reference rather than value, it can be a bit unclear if that method considers the space used by that data structure when calculating its space complexity. If you didn't count it, then it would be easy for all our methods to have great space usage on paper because we put the onus on the caller to allocate that space. If we did count it, but the data structure was created for use by many different methods then is the space complexity for all those methods O(N) when they aren't utilising additional space. Then consider that if your method receives an array as an input and loops it, then an index must be created for the loop which uses additional space.
+
+The first answer [to this question](https://cs.stackexchange.com/questions/127933/analyzing-space-complexity-of-passing-data-to-function-by-reference) provides some great context to the question and gives some thought provoking answers.
```

Comment:
The link text should be updated. "To this question" won't convey any useful context when isolated.

```suggestion
The first answer to [analyzing space complexity](https://cs.stackexchange.com/questions/127933/analyzing-space-complexity-of-passing-data-to-function-by-reference) provides some great context to the question and gives some thought provoking answers.
```